Add helper method for accessing contents of typedarrays

mikolalysenko opened the pull request:
This PR adds a portable helper method for reading the contents of a typedarray (ArrayBufferView) from C++, fixing https://github.com/nodejs/nan/issues/455.

I tried to follow the API conventions from `Utf8String` for this module closely.  Note that this does not make a copy of the typedarray data or hold a reference to the underlying array object, so it would require the user to know what they are doing and not try to access these bytes outside the natural lifetime of the object.

bnoordhuis commented on nan_typedarray_contents.h:
```diff
@@ -0,0 +1,82 @@
+// Copyright Joyent, Inc. and other Node contributors.
```
Why (c) Joyent?

mikolalysenko replied:
I just copy-pasted the boilerplate from nan.h.  If you have an alternative license I should use, let me know.

bnoordhuis replied:
Did you mean nan_string_bytes.h?  I don't see Joyent copyright boilerplate anywhere else.

Aside, I speculate @indutny included it because it was copied from joyent/node but I think we can just remove it; most of that code was originally written by me.

indutny replied:
I'm fine with removing this.

mikolalysenko replied:
Yes, I just grabbed a random file and must've picked the wrong one.

bnoordhuis commented on nan_typedarray_contents.h:
```diff
@@ -0,0 +1,82 @@
+        v8::Local<v8::ArrayBufferView>::Cast(from);
+
+      int byteLength = array->ByteLength();
+      int byteOffset = array->ByteOffset();
```
Can you use `size_t` here and in general?  Maybe make them `const` while you're here.

mikolalysenko replied:
Sure.

bnoordhuis commented on nan_typedarray_contents.h:
```diff
@@ -0,0 +1,82 @@
+#if defined(V8_MAJOR_VERSION) && (V8_MAJOR_VERSION > 4 ||                      \
+  (V8_MAJOR_VERSION == 4 && defined(V8_MINOR_VERSION) && V8_MINOR_VERSION >= 3))
+
+    if(from->IsArrayBufferView()) {
```
Minor style infraction: space after `if`.

kkoopa commented on test/cpp/typedarrays.cpp:
```diff
@@ -0,0 +1,68 @@
+
+  v8::Local<v8::Array> result = Nan::New<v8::Array>(data.length());
+  for (int i=0; i<data.length(); ++i) {
+    result->Set(i, Nan::New<v8::Number>((*data)[i]));
```
`Nan::Set(result, ...`

bnoordhuis commented on nan_typedarray_contents.h:
```diff
@@ -0,0 +1,82 @@
+
+      void* data = buffer->GetContents().Data();
+      length_ = byteLength / sizeof(T);
+      data_   = (T*)((void*)((char*)data + byteOffset));
```
This will result in undefined behavior if the pointer's alignment doesn't satisfy `alignas(T)`.

Aside, can you use C++ casts here?

mikolalysenko replied:
Using C++ casts is no big deal.  Regarding alignment, what do you think the right behavior should be?  I can think of at least 3 options:

1.  Do nothing.
2.  Detect bad alignment at run time and return NULL.
3.  Make a local copy with the correct alignment.

mikolalysenko replied:
Actually there is a 4th option:

* Use `std::align` to coerce the pointer to an aligned value.

kkoopa commented on test/cpp/typedarrays.cpp:
```diff
@@ -0,0 +1,68 @@
+
+  v8::Local<v8::Array> result = Nan::New<v8::Array>(data.length());
+  for (int i=0; i<data.length(); ++i) {
+    result->Set(i, Nan::New<v8::Number>((*data)[i]));
```
`Nan::Set`

bnoordhuis commented on nan_typedarray_contents.h:
```diff
@@ -0,0 +1,82 @@
+      v8::Local<v8::Value> byteOffset = array->Get(
+        Nan::New<v8::String>("byteOffset").ToLocalChecked());
+
+      if(byteLength->IsUint32() && byteOffset->IsUint32()) {
```
Style error.

kkoopa commented on test/cpp/typedarrays.cpp:
```diff
@@ -0,0 +1,68 @@
+
+  v8::Local<v8::Array> result = Nan::New<v8::Array>(data.length());
+  for (int i=0; i<data.length(); ++i) {
+    result->Set(i, Nan::New<v8::Number>((*data)[i]));
```
`Nan::Set`

bnoordhuis commented on nan_typedarray_contents.h:
```diff
@@ -0,0 +1,82 @@
+
+      if(byteLength->IsUint32() && byteOffset->IsUint32()) {
+
+        unsigned int length = byteLength->Uint32Value();
```
`const uint32_t`?

kkoopa commented on test/cpp/typedarrays.cpp:
```diff
@@ -0,0 +1,68 @@
+
+  v8::Local<v8::Array> result = Nan::New<v8::Array>(data.length());
+  for (int i=0; i<data.length(); ++i) {
+    result->Set(i, Nan::New<v8::Number>((*data)[i]));
```
`Nan::Set`

bnoordhuis commented on nan_typedarray_contents.h:
```diff
@@ -0,0 +1,82 @@
+        unsigned int length = byteLength->Uint32Value();
+        void* data = array->GetIndexedPropertiesExternalArrayData();
+
+        if(data) {
```
Style error.  I'm going to stop pointing them out from now on.

bnoordhuis commented on nan_typedarray_contents.h:
```diff
@@ -0,0 +1,82 @@
+
+ private:
+   int  length_;
+   T*   data_;
```
Consider making these `const` (`T* const` in the case of `data_`.)

bnoordhuis commented on test/cpp/typedarrays.cpp:
```diff
@@ -0,0 +1,68 @@
+  Nan::TypedArrayContents<uint8_t> data(info[0]);
+
+  v8::Local<v8::Array> result = Nan::New<v8::Array>(data.length());
+  for (int i=0; i<data.length(); ++i) {
```
Style errors: needs more spaces.

kkoopa commented:
Would it make sense to make the class non-heap-allocatable? or is there a legitimate use case for having it on the heap, what with the lifetime issues and that?

mikolalysenko commented:
I agree that there doesn't seem to be a good reason for heap ownership, so I think it seems reasonable to disable it.  If you did need to track the data on the heap anyway, it would be no big deal to instantiate the class locally and copy the data over manually.

mikolalysenko commented:
Ok, I fixed up some of the style issues and switched some of the types and macros around to be more idiomatic/descriptive.

I still haven't done anything with alignment, since this requires making a decision on what to do in the case of unaligned data.  I think that doing nothing is reasonable, but if we commit to this, then it should be documented in `v8_misc.md`.

indutny commented:
Just curious, what are you going to do about alignment?

mikolalysenko commented:
@indutny currently leaning toward doing nothing, but returning NULL seems ok too.

indutny commented:
Aah, I re-read the comments. Everything is clear now, nvm.

bnoordhuis commented on nan_typedarray_contents.h:
```diff
@@ -0,0 +1,77 @@
+      const ptrdiff_t byteOffset = array->ByteOffset();
+      v8::Local<v8::ArrayBuffer> buffer = array->Buffer();
+
+      char* data = reinterpret_cast<char*>(buffer->GetContents().Data());
```
Can you use `static_cast` here?

bnoordhuis commented on nan_typedarray_contents.h:
```diff
@@ -0,0 +1,77 @@
+        v8::Local<v8::ArrayBufferView>::Cast(from);
+
+      const size_t    byteLength = array->ByteLength();
+      const ptrdiff_t byteOffset = array->ByteOffset();
```
Minor style nit but I think nan predominantly uses snake_case, not camelCase, for local variables.

bnoordhuis commented on nan_typedarray_contents.h:
```diff
@@ -0,0 +1,77 @@
+      char* data = reinterpret_cast<char*>(buffer->GetContents().Data());
+
+      length_ = byteLength / sizeof(T);
+      data_   = reinterpret_cast<T*>(data + byteOffset);
```
I think I'd insert an assert that checks the alignment.  You could use `alignof(T)` here but not in general because it's a C++11 construct (io.js v3 and node v4 imply C++11 support but that's not the case with node v0.10 and v0.12.)

Maybe settle for `assert(reinterpret_cast<uintptr_t>(data + byte_offset) % sizeof(T) == 0)`.  There are perhaps still one or two edge cases that will slip through, though.  I'm thinking of things like performing SSE operations on an array of doubles that only has an alignment of 8.

mikolalysenko replied:
What about situations where the user would read out the contents of a typed array and then immediately copy it to local storage?  For example, maybe you want to read out the typedarray data and then immediately pass it to the OpenGL driver, which internally just copies it out.

bnoordhuis replied:
Then you'd just use a `TypedArrayContents<char>`, wouldn't you?

mikolalysenko replied:
Fair enough.

bnoordhuis commented on test/cpp/typedarrays.cpp:
```diff
@@ -0,0 +1,68 @@
+  Nan::TypedArrayContents<int32_t> data(info[0]);
+
+  v8::Local<v8::Array> result = Nan::New<v8::Array>(data.length());
+  for (size_t i=0; i<data.length(); ++i) {
```
Can you fix the style issues in the test?

mikolalysenko commented:
Ok, I added an assert to check the alignment and made a few more style changes in the test cases.

mikolalysenko commented:
Ah whoops, I messed up the rebase and named that last commit incorrectly.  It should say "assert false when typedarrays missing" or something to that effect.

kkoopa commented:
[AppVeyor says no-go](https://ci.appveyor.com/project/RodVagg/nan/build/1007/job/uvctrj09asbd5qyq#L242).

```
c:\projects\nan\nan_typedarray_contents.h(67): error C2275: 'T' : illegal use of this type as an expression [C:\projects\nan\test\build\typedarrays.vcxproj]
          ..\cpp\typedarrays.cpp(16) : see declaration of 'T'
          c:\projects\nan\nan_typedarray_contents.h(15) : while compiling class template member function 'Nan::TypedArrayContents<uint8_t>::TypedArrayContents(v8::Local<v8::Value>)'
          ..\cpp\typedarrays.cpp(16) : see reference to function template instantiation 'Nan::TypedArrayContents<uint8_t>::TypedArrayContents(v8::Local<v8::Value>)' being compiled
          ..\cpp\typedarrays.cpp(16) : see reference to class template instantiation 'Nan::TypedArrayContents<uint8_t>' being compiled
c:\projects\nan\nan_typedarray_contents.h(67): error C3861: 'alignof': identifier not found [C:\projects\nan\test\build\typedarrays.vcxproj]
```

mikolalysenko commented:
Ok, I took out `alignof` since it seems that doesn't work on windows.  We'll see if the new version passes.

kkoopa commented:
Windows should have `__alignof`.

mikolalysenko commented:
Alright, I added in some more precise alignments checks for various platforms.

mikolalysenko commented:
@kkoopa can you start the appveyor CI?  It looks like the node v0.12 case failed due to a network problem during the npm install.

kkoopa commented:
Can't. Push something to make it run.

On September 15, 2015 8:03:08 PM GMT+03:00, Mikola Lysenko <notifications@github.com> wrote:
>@kkoopa can you start the appveyor CI?  It looks like the node v0.12
>case failed due to a network problem during the npm install.
>
>---
>Reply to this email directly or view it on GitHub:
>https://github.com/nodejs/nan/pull/459#issuecomment-140463526

mikolalysenko commented:
Ok, I pushed a doc tweak and it passed this time.

bnoordhuis commented on nan_typedarray_contents.h:
```diff
@@ -0,0 +1,89 @@
+    assert(reinterpret_cast<uintptr_t>(data_) % alignof(T) == 0);
+#else
+    assert(reinterpret_cast<uintptr_t>(data_) % sizeof(T) == 0);
+#endif
```
I'd hoist this to the top of the constructor and replace the NODE_MODULE_VERSION check with `#if __cplusplus >= >= 201103L` (and make it the first one.)

PS: Both clang and gcc have `__alignof(...)` so you should be able to fold the gcc and VS code paths.

mikolalysenko replied:
If we hoist it to the top of the constructor, then `data_` won't be initialized.  It is probably better to check the assert after we computed the byteOffset and finished all the unboxing logic.

bnoordhuis replied:
Yes, that's what I mean; the assert should precede assigning to `data_`.

mikolalysenko replied:
So, what you would want is to maybe read data_ and length_ into temporary variables, then write the result only if the assertion passes?

mikolalysenko commented:
Is this PR ok?  I would like to use it in headless-gl so that it can support node v4.  If the commit log is too messy I can do a rebase and squish it into a new PR.

kkoopa commented:
Would it not be more appropriate to have it return a `Maybe<T>` instead of possibly returning `NULL`. It seems a better fit with the rest that way.

mikolalysenko commented:
@kkoopa I'm not sure what the benefit of wrapping everything in a `Maybe<>` would be, besides making things more complicated.  `Utf8String` doesn't do this, and I think there is a strong parallel between that usage and the situation here.

kkoopa commented:
The difference is that Utf8String cannot return `NULL`. The point of wrapping everything in Maybes is to get rid of `NULL`.

mikolalysenko commented:
Well, why stop there.  We could even add some overloads for `const` and all permutations of `Maybe<>` as an argument:

```c++
template<typename T> Nan::Maybe< Nan::TypedArrayContents<T> > getTypedArrayContents(Nan::Maybe< Nan::Local<Value> > &object);

template<typename T> Nan::Maybe< const Nan::TypedArrayContents<T> > getTypedArrayContents(const Nan::Maybe< Nan::Local<Value> > &object);

template<typename T> Nan::Maybe< Nan::TypedArrayContents<T> > getTypedArrayContents(Nan::Local<Value> &object);

template<typename T> Nan::Maybe< const Nan::TypedArrayContents<T> > getTypedArrayContents(const Nan::Local<Value> &object);
```

mikolalysenko commented:
Alternatively, we could maybe use a different interface.  Perhaps the following approach would be simpler:

```c++
template<typename T>
  bool Nan::getTypedArrayContents(Nan::Local<Value> array, size_t& length, T*& data);
```

Where this method returns `false` if the object is not a valid typedarray.

jfirebaugh commented:
@kkoopa Seems to me that `Utf8String` does provide a precent for using NULL. From [the docs](https://github.com/nodejs/nan/blob/master/doc/v8_misc.md#nanutf8string):

> If conversion to a string fails (e.g. due to an exception in the toString() method of the object) then the length() method returns 0 and the * operator returns NULL.

I'm :+1: on merging this in the current state.

mikolalysenko commented:
I'm ok with changing the interface too, just let me know how you want it to look.

kkoopa commented:
@jfirebaugh Good point, although the documentation might not be accurate here. At least I don't remember enforcing it anywhere: https://github.com/nodejs/nan/blob/master/nan.h#L861-L901

Who knows when V8 will change `Utf8Value` to return a `Maybe`? Seems like it should be changed, to make it consistent with the new ways, but that is a different matter.

I don't have particularly strong feelings either way, it's mostly caution with what will be the best way of introducing a new API and in what form, as there will be no option for take-backs until the next major version.

Perhaps @bnoordhuis could weigh in on this?

mikolalysenko commented:
Hold the press on this one, I found a failing case.  On node v0.12, it seems that there is some optimization where typed arrays aren't allocated until they are written to.  I'm not sure what is causing this problem, but it is not present in earlier versions and does not fail in node v4.

mikolalysenko commented:
Figured out a workaround.

mikolalysenko commented:
I think the CI failed due to a network issue.  If you want I can poke this PR with some doc tweak to rerun it.

kkoopa commented:
Poke ahead. There are occasional network issues.

On September 23, 2015 9:01:59 PM GMT+03:00, Mikola Lysenko <notifications@github.com> wrote:
>I think the CI failed due to a network issue.  If you want I can poke
>this PR with some doc tweak to rerun it.
>
>---
>Reply to this email directly or view it on GitHub:
>https://github.com/nodejs/nan/pull/459#issuecomment-142681178

mikolalysenko commented:
I'm not sure what is causing the problems on node v0.8, as far as I know nothing should have changed.  It looks like travis is failing to install the module `are-we-there-yet`.  Locally, I can run the tests against 0.8 and everything passes with no problems.

kkoopa commented:
Could it be that the travis config does not update npm for node 0.8? If it works locally it is probably good... 0.8 can act flaky with npm sometimes. I've seen that before I think.

On September 24, 2015 12:05:25 AM GMT+03:00, Mikola Lysenko <notifications@github.com> wrote:
>I'm not sure what is causing the problems on node v0.8, as far as I
>know nothing should have changed.  It looks like travis is failing to
>install the module `are-we-there-yet`.  Locally, I can run the tests
>against 0.8 and everything passes with no problems.
>
>---
>Reply to this email directly or view it on GitHub:
>https://github.com/nodejs/nan/pull/459#issuecomment-142727774

mikolalysenko commented:
Yeah, it is just node v0.8 on CI.  Everything works fine locally.

LinusU commented:
@mikolalysenko If you want to trigger ci without adding more commits you can change the latest one instead.

```sh
git commit --amend --no-edit
```

Then force push to overwrite it

```sh
git push --force
```

Just a small tip :)

kkoopa commented:
Seems npm has been updated to 3.3 and they broke that damn are-we-there-yet thing, *again*. Why can't they stop touching it or simply remove it completely?

bsergean commented:
Should an issue be created in are-we-there-yet, or in npm ?

https://github.com/iarna/are-we-there-yet/issues

tmpvar commented:
:+1:, what can I do to help get this landed?

rvagg commented:
https://github.com/npm/npm/search?q=are-we-there-yet&type=Issues&utf8=%E2%9C%93 - there's plenty of issues & discussion about the npm problem

mikolalysenko commented:
I removed the extra const qualifiers on the value types returned by the accessors, since they aren't needed and they also generate warnings when compiling with GCC on Linux.

kkoopa commented:
It should have both versions of the operators. Both the const- and non-const versions.

mikolalysenko commented:
The issue is not making the contents of the pointer const, but rather making the pointer itself const.  Returning a const pointer isn't needed since you can freely cast it to a non-const pointer.  While doing this would have no semantic effect, it does trigger a warning in gcc on Linux (but not in clang on OS X) which you would have to ignore.

Forcing users to ignore warnings seems like it could cause problems, especially if you are building with -Werror.

kkoopa commented:
Ah, yes. I did not read the diff thoroughly.

kkoopa commented:
If you rebase and squash, I guess it is good to merge.

mikolalysenko commented:
Closing this PR, moved all the changes into a separate branch.
